### PR TITLE
Allow Intersection Types in GraphQL Services

### PR DESCRIPTION
**Pull request by ThisaruGuruge**
## Purpose

With this PR, intersection types are allowed as input parameters in Ballerina GraphQL resolver functions.

Fixes: [#2620](https://github.com/ballerina-platform/ballerina-standard-library/issues/2620)

## Examples

Refer to the linked issue

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: compiler-plugin-tests/src/test/java/io/ballerina/stdlib/graphql/compiler/CompilerPluginTest.java [hunk 1]
```diff
@@ -144,12 +144,19 @@ public void testGraphQLListTypeInputParametersInResources() {
...
         Assert.assertEquals(diagnosticResult.errorCount(), 0);
     }
 
+    @Test
+    public void testIntersectionTypes() {
+        String packagePath = "valid_service_16";
+        DiagnosticResult diagnosticResult = getDiagnosticResult(packagePath);
+        Assert.assertEquals(diagnosticResult.errorCount(), 0);
+    }
+
```
Comment (DimuthuMadushan) [inline compiler-plugin-tests/src/test/java/io/ballerina/stdlib/graphql/compiler/CompilerPluginTest.java:159]: Shall we add a test for a service using the same intersection type as input and return type?

File: ballerina/tests/resources/documents/intersection_input.graphql [hunk 2]
```diff
@@ -0,0 +1,55 @@
+query getName {
+    name(species: { genus: "Craspedocephalus", specificName: "trigonocephalus" })
```
Comment (DimuthuMadushan) [inline ballerina/tests/resources/documents/intersection_input.graphql:2]:
```suggestion
    name(species: {genus: "Craspedocephalus", specificName: "trigonocephalus"})
```

File: ballerina/tests/resources/documents/intersection_input.graphql [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+query getCity {
+    city(address: { number: "308", street: "Negra Arroyo Lane", city: "Albuquerque" })
```
Comment (DimuthuMadushan) [inline ballerina/tests/resources/documents/intersection_input.graphql:6]:
```suggestion
    city(address: {number: "308", street: "Negra Arroyo Lane", city: "Albuquerque"})
```

File: ballerina/tests/resources/documents/intersection_input.graphql [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            genus: "Kallima",
+            specificName: "philarchus"
+        }
+    ])
+}
+
+query getCities {
+    cities(addresses: [
+        { number: "308", street: "Negra Arroyo Lane", city: "Albuquerque" },
+        { number: "9809", street: "Margo Street", city: "Albuquerque" }
```
Comment (DimuthuMadushan) [inline ballerina/tests/resources/documents/intersection_input.graphql:39]: ditto

File: compiler-plugin-tests/src/test/java/io/ballerina/stdlib/graphql/compiler/CompilerPluginTest.java (hunk 1 above)
Comment (ThisaruGuruge) [reply compiler-plugin-tests/src/test/java/io/ballerina/stdlib/graphql/compiler/CompilerPluginTest.java:159]: I thought the use of input as an output (and vice-versa) is handled by the specific test cases. Do we need separate tests for that in this scenario?

Comment (DimuthuMadushan) [reply compiler-plugin-tests/src/test/java/io/ballerina/stdlib/graphql/compiler/CompilerPluginTest.java:159]: In those test cases, we didn't use intersection types as I remember.

Comment (ThisaruGuruge) [reply compiler-plugin-tests/src/test/java/io/ballerina/stdlib/graphql/compiler/CompilerPluginTest.java:159]: Yes, but the point is, input-output type checking is done separately, without considering whether it is an intersection type or not. So, if the same type is used as an input and as an output, it should throw an error regardless of it being an intersection type, right?

File: ballerina/tests/resources/documents/intersection_types.graphql [hunk 5]
```diff
@@ -0,0 +1,77 @@
+query getName {
+  name(species: { genus: "Craspedocephalus", specificName: "trigonocephalus" })
```
Comment (DimuthuMadushan) [inline ballerina/tests/resources/documents/intersection_types.graphql:2]:
```suggestion
  name(species: {genus: "Craspedocephalus", specificName: "trigonocephalus"})
```
is this the correct formatting here? Seems like starting and ending spaces are added automatically.

Comment (ThisaruGuruge) [reply ballerina/tests/resources/documents/intersection_types.graphql:2]: Yes, the GraphQL formatter adds the spaces.

**Review by DimuthuMadushan — APPROVED** [review]

**Pull request merged**